### PR TITLE
Editoast: add new simulation endpoint

### LINKED ISSUE #6853: TS2: Add editoast simulation endpoint
### Description and goal

Implementation of the `GET` `/v2/train_schedule/ID/simulation?infra=N` endpoint.

### Acceptance criteria
The endpoint structure is already implemented. 

The endpoint should retrieve: 
- [ ] The same output as Core endpoint output (https://github.com/osrd-project/osrd/issues/6855)
- [ ] A 404 if an infra or a train schedule is not found 

To do:
- [ ] Compute the hash of the input and output and cache them
- [ ] Setup core client to call core endpoint

### PR DESCRIPTION
**Pull request by younesschrifi**
close #6853

### CODE REVIEW

File: editoast/editoast_derive/src/custom_hash.rs
Comment (woshilapin) [inline editoast/editoast_derive/src/custom_hash.rs:1]: I'm not sure we really need this custom macro. It adds maintenance complexity, and we already have `derivative` as a dependency that, I believe, should be enough to do the job. See the `#[derivative(Hash(hash_with="path::to::my_hash_fn"))]` in [here](https://mcarton.github.io/rust-derivative/latest/Hash.html#hash-with).

Also, I know we don't need `Eq` at the moment, but I'm worried that in the future, we add a `#[derive(PartialEq, Eq)]` and do not respect [this contract](https://doc.rust-lang.org/std/hash/trait.Hash.html#hash-and-eq), which might cause hard to debug logic bugs (note: I don't know if it's any better with `derivative` by the way).

File: editoast/src/views/v2/path/pathfinding.rs [hunk 1]
```diff
@@ -53,11 +53,18 @@ crate::schemas! {
...
         blocks: Vec<Identifier>,
         #[schema(inline)]
+        /// Path description as route ids
         routes: Vec<Identifier>,
+        /// Path description as track ranges
         track_section_ranges: Vec<TrackRange>,
+        /// Length of the path in mm
         length: u64,
+        /// The path offset of each path item given as input of the pathfinding
+        /// The first value is always `0` (begining of the path) and the last one is always equal to the `length`
```
Comment (woshilapin) [inline editoast/src/views/v2/path/pathfinding.rs]:
```suggestion
        /// The first value is always `0` (beginning of the path) and the last one is always equal to the `length`
```

Comment (woshilapin) [inline editoast/src/views/v2/path/pathfinding.rs]: I'm not sure I understand what this field is for and the documentation doesn't help me. Especially, with the last sentence, I understand that the vector is filled with numbers from 0 to length and surely, it's not very useful to send a vector of all integers up to a number.

File: front/public/locales/en/errors.json [hunk 2]
```diff
@@ -170,7 +170,8 @@
     },
     "train_schedule_v2": {
       "BatchTrainScheduleNotFound": "'{{number}}' train schedule(s) could not be found",
-      "NotFound": "Train Schedule '{{train_schedule_id}}' could not be found"
+      "NotFound": "Train Schedule '{{train_schedule_id}}' could not be found",
+      "InfraNotFound": "Infra '{{infra_id}} could not be found"
```
Comment (woshilapin) [inline front/public/locales/en/errors.json]:
```suggestion
      "InfraNotFound": "Infra '{{infra_id}}' could not be found"
```

File: front/public/locales/fr/errors.json [hunk 3]
```diff
@@ -117,8 +117,8 @@
     },
     "scenario": {
       "NotFound": "Scénario non trouvé",
-      "InfraNotFound":"Infrastructure {{infra_id}} non trouvée",
-      "TimetableNotFound":"Grille horaire '{{timetable_id}}' non trouvée"
+      "InfraNotFound": "Infrastructure {{infra_id}} non trouvée",
```
Comment (woshilapin) [inline front/public/locales/fr/errors.json:120]:
```suggestion
      "InfraNotFound": "Infrastructure '{{infra_id}}' non trouvée",
```

File: front/public/locales/fr/errors.json [hunk 4]
```diff
@@ -172,7 +172,8 @@
     },
     "train_schedule_v2": {
       "BatchTrainScheduleNotFound": "'{{number}}' circulation(s) n'ont pas pu être trouvée(s)",
-      "NotFound": "Circulation '{{train_schedule_id}}' non trouvée"
+      "NotFound": "Circulation '{{train_schedule_id}}' non trouvée",
+      "InfraNotFound": "Infra '{{infra_id}} non trouvée"
```
Comment (woshilapin) [inline front/public/locales/fr/errors.json]:
```suggestion
      "InfraNotFound": "Infra '{{infra_id}}' non trouvée"
```

File: front/public/locales/fr/errors.json [hunk 5]
```diff
@@ -117,8 +117,8 @@
     },
     "scenario": {
       "NotFound": "Scénario non trouvé",
-      "InfraNotFound":"Infrastructure {{infra_id}} non trouvée",
-      "TimetableNotFound":"Grille horaire '{{timetable_id}}' non trouvée"
```
Comment (woshilapin) [inline front/public/locales/fr/errors.json:121]: Is it intended that these 2 keys are not translated in the `en` version (I know it's not related to this PR, but this change just made me check). Feel free to discard and/or send it to someone else for an answer/fix.

File: editoast/src/views/v2/train_schedule.rs [hunk 6]
```diff
@@ -255,8 +268,18 @@ async fn put(
     Ok(Json(ts_result.into()))
 }
 
-#[derive(Deserialize, Serialize, Clone, Debug, ToSchema)]
-struct SimulationOutput {
+// TODO: When can the simulation fail ?
+#[derive(Serialize, Deserialize, Clone, Debug, ToSchema)]
+#[serde(tag = "status", rename_all = "snake_case")]
+enum SimulationResult {
+    Success(Box<SimulationResultSuccess>),
```
Comment (woshilapin) [inline editoast/src/views/v2/train_schedule.rs]: Why do we need to `Box` here? Is it because the enum's variants have very different size and `clippy` is complaining about it ([like this](https://rust-lang.github.io/rust-clippy/master/index.html#/large_enum_variant))? If I understand correctly, the `PathFindingFailed` variant is for an error case, so maybe we don't care too much to run in a suboptimal way (a lot of memory wasted in this case) for an error case? Maybe a `#[allow(clippy::large_enum_variant)]` would be ok here? I'm really exploring the idea here, I'm not sure myself, happy to take feedback on this idea.

File: editoast/src/views/v2/train_schedule.rs [hunk 7]
```diff
@@ -318,75 +370,252 @@ pub struct MrspPoint {
...
+    let rolling_stock = RollingStockModel::retrieve(conn, rolling_stock_name.clone())
+        .await?
+        .expect("Rolling stock should exist since the pathfinding succeeded");
+    // Get electrical_profile_set_id
+    let timetable_id = train_schedule.timetable_id;
+    let timetable = Timetable::retrieve(conn, timetable_id)
+        .await?
+        .expect("Timetable should exist since it's a foreign key");
+
+    assert_eq!(path_items_position.len(), train_schedule.path.len());
```
Comment (woshilapin) [inline editoast/src/views/v2/train_schedule.rs:544]: Instead of a `panic`, since we're in a fallible function, why not returning an `Err`?

File: editoast/src/views/v2/train_schedule.rs [hunk 8]
```diff
@@ -318,75 +370,252 @@ pub struct MrspPoint {
...
+
+    // Project path items to path offset
+    let path_items_to_position: HashMap<_, _> = train_schedule
+        .path
+        .iter()
+        .map(|p| &p.id)
+        .zip(path_items_position.iter())
```
Comment (woshilapin) [inline editoast/src/views/v2/train_schedule.rs]: It's probably not really that useful to have `&u64` as values, let's make it `u64` directly.

```suggestion
        .zip(path_items_position.iter().copied())
```

File: editoast/src/views/v2/train_schedule.rs [hunk 9]
```diff
@@ -318,75 +370,252 @@ pub struct MrspPoint {
...
+
+    let schedule = train_schedule
+        .schedule
+        .iter()
+        .map(|schedule_item| SimulationScheduleItem {
+            path_offset: *path_items_to_position[&schedule_item.at],
+            arrival: schedule_item
+                .arrival
+                .as_ref()
+                .map(|t| t.num_seconds() as u64),
```
Comment (woshilapin) [inline editoast/src/views/v2/train_schedule.rs:562]: This can fail if `Duration` is negative (not sure that it panics, by the way, it probably does some weird things). Should we have a `PositiveDuration` somewhere? It might a be a job for another PR though.

**Review by woshilapin — COMMENTED** [review]
Thanks for adding this new (big!) endpoint. I made a few comments.

I question the custom macro `CustomHash` for a few different reasons.

Also, the PR is quite big, maybe it could have been split in different pieces to ease reviewing?

File: editoast/src/views/v2/train_schedule.rs (hunk 9 above)
Comment (flomonster) [reply editoast/src/views/v2/train_schedule.rs:562]: We can actually adapt our custom `Duration` into a `PositiveDuration`

File: editoast/src/views/v2/path/pathfinding.rs (hunk 1 above)
Comment (flomonster) [reply editoast/src/views/v2/path/pathfinding.rs]: We'll try to rephrase that.

------------

FYI: This field returns the position relative to the waypoint path given as input.

### Example 1:

If we have a train schedule with only two waypoints (starting and ending point). And the computed path is 1km long:
Then the response will be `path_items_positions: [0, 1_000_000]`

### Example 2:

If we have a train schedule with four waypoints (`A`, `B`, `C` and `D`). And the computed path is 1km long:
Then a possible response is `path_items_positions: [0, 250_000, 600_000, 1_000_000]`
This means that `B` is 250m from the start of the path and `C` is 600m away.

File: editoast/editoast_derive/src/custom_hash.rs
Comment (flomonster) [reply editoast/editoast_derive/src/custom_hash.rs:1]: If use `derivative` crate we have to make multiple hash functions for each desired rounding level. I still think that it's a better idea and easier to maintain than `CustomHash`.

I don't think we have a problem implementing `Eq`. We're not breaking the contract.
Taking `a` and `b` of type `f64` if `a == b` then the `hash(a.round()) == hash(b.round())`

File: editoast/src/views/v2/path/pathfinding.rs (hunk 1 above)
Comment (woshilapin) [reply editoast/src/views/v2/path/pathfinding.rs]: Oh, then length is not the length of a vec, but the length of a path, indeed, should have thought about that. Maybe `Vec::len()` did trigger me in the wrong way.

To help rephrasing it, maybe you could put the unit in the sentence, because I basically understood indexes were in it.

File: editoast/editoast_derive/src/custom_hash.rs
Comment (woshilapin) [reply editoast/editoast_derive/src/custom_hash.rs:1]: Thanks for the feedback, indeed, I took the `a == b => hash(a) == hash(b)` the wrong way 🤦 Nice if you can make it work with `derivative`.

File: front/public/locales/fr/errors.json (hunk 5 above)
Comment (flomonster) [reply front/public/locales/fr/errors.json:121]: We've removed the two errors (on the editoast side) and they weren't translated into English. I think someone forgot to do it on a previous PR.

File: editoast/src/core/v2/simulation.rs [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    pub mass: u64,
+    pub rolling_resistance: RollingResistance,
+    /// Mapping of power restriction code to power class
+    #[serde(default)]
+    pub power_restrictions: BTreeMap<String, String>,
+    /// The time the train takes before actually using electrical power (in miliseconds). Is null if the train is not electric.
+    pub electrical_power_startup_time: Option<u64>,
+    /// The time it takes to raise this train's pantograph in seconds. Is null if the train is not electric.
+    /// The time is in miliseconds.
+    pub raise_pantograph_time: Option<u64>,
```
Comment (woshilapin) [inline editoast/src/core/v2/simulation.rs]: How about grouping these 2 properties into one `struct` (maybe with `#[serde(flatten)]`)? That would mean only one `Option` to handle in the code for when it's electric.

File: editoast/src/schema/utils/duration.rs [hunk 11]
```diff
@@ -50,29 +57,36 @@ use serde::Serialize;
...
+    /// This function errors when the given duration is negative
+    /// The created PositiveDuration is limited to 1 millisecond
+    pub fn from_chrono(duration: ChronoDuration) -> Result<Self, PositiveDurationError> {
+        let milli_sec = duration.num_milliseconds();
+        if milli_sec < 0 {
+            return Err(PositiveDurationError::NegativeDuration);
+        }
+        Ok(PositiveDuration(ChronoDuration::milliseconds(milli_sec)))
     }
 }
```
Comment (woshilapin) [inline editoast/src/schema/utils/duration.rs]: Why not implementing `TryFrom<ChronoDuration>` instead of a free function?

File: editoast/src/views/v2/train_schedule.rs [hunk 12]
```diff
@@ -255,37 +267,47 @@ async fn put(
...
-    provisional: ReportTrain,
-    final_output: CompleteReportTrain,
-    mrsp: Mrsp,
-    power_restriction: String,
+// TODO: When can the simulation fail ?
+#[derive(Serialize, Deserialize, Clone, Debug, ToSchema)]
+#[serde(tag = "status", rename_all = "snake_case")]
+#[allow(clippy::large_enum_variant)]
+enum SimulationResult {
+    #[allow(clippy::large_enum_variant)]
```
Comment (woshilapin) [inline editoast/src/views/v2/train_schedule.rs]: Maybe give that `#[allow(...)]` a comment.

Suggestion: The other variant is only used when an error occurs, it's preferable to avoid indirection from a `Box`on a happy path, and it's ok to waste some memory in an error path.

File: editoast/src/views/v2/train_schedule.rs (hunk 9 above)
Comment (woshilapin) [reply editoast/src/views/v2/train_schedule.rs:562]: Seems resolved from what I can see. Thank you 😍

File: editoast/editoast_derive/src/custom_hash.rs
Comment (woshilapin) [reply editoast/editoast_derive/src/custom_hash.rs:1]: Nice work, seems resolved.

**Review by woshilapin — APPROVED** [review]

**Review by flomonster — APPROVED** [review]

File: editoast/src/core/v2/simulation.rs (hunk 10 above)
Comment (younesschrifi) [reply editoast/src/core/v2/simulation.rs]: One field can be null while the other is not, we've changed the doc.

**Review by clarani — APPROVED** [review]
LGTM :white_check_mark:
